### PR TITLE
[ADD] web_editor: two new powerbox widgets to rate with 3 or 5 stars

### PR DESCRIPTION
**Pull request by Zynton**
This adds two new widgets to the powerbox, which allow the user to quickly insert 3 or 5 stars, then click on them to color them gold.
Similar to checklists, this allows the user to click the powerbox stars widgets even in readonly mode, so as to change a rating without going in edit mode.

While implementing this it became apparent that the checklist readonly checking feature was broken so this fixes it too. In order to be able to import Odoo Editor utils into field_html, the widget was turned into a js module with modern imports.

task-2575449

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Zinston** [comment]
@odoo/rd-security The trigger is just a call to an editor command called `insertHTML`, so as to insert hard coded HTML into the editable area.

PS: the runbot test that is red is the same as in master, nothing to do with my branch.

**Comment by mart-e** [comment]
@robodoo override=ci/security
Reason: inserting static html content

**Comment by Zinston** [comment]
Just rebased now that master is green so it should all be ok.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 1]
```diff
@@ -2524,6 +2550,28 @@ export class OdooEditor extends EventTarget {
                 this.historyStep();
             }
         }
+
+        // handle stars
+        if (node.nodeType === Node.ELEMENT_NODE && node.className.includes('fa-star')) {
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2562]: Shouldn't this code be protected to only affect your new snippets ? Or maybe I'm missing something ?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 2]
```diff
@@ -1510,22 +1515,7 @@ const Wysiwyg = Widget.extend({
...
-            self.trigger_up('wysiwyg_attachment', attachments);
-        };
-        options.onFocus = function () {
-            self.trigger_up('wysiwyg_focus');
-        };
-        options.onBlur = function () {
-            self.trigger_up('wysiwyg_blur');
-        };
-        return options;
+        return Object.assign({}, this.defaultOptions, this.options);
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1520]: Why this change in this commit ?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 3]
```diff
@@ -445,6 +445,7 @@ export class OdooEditor extends EventTarget {
             this.observer = new MutationObserver(records => {
                 records = this.filterMutationRecords(records);
                 if (!records.length) return;
+                this.dispatchEvent(new Event('contentChanged'));
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:448]: I'm not sure I understand the link between this change and the checklist feature ?

File: addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js [hunk 4]
```diff
@@ -133,6 +133,11 @@ class Sanitize {
             }
         }
 
+        // Ensure unique ids on checklists.
+        if (node.nodeName == 'LI' && node.parentElement && node.parentElement.classList.contains('o_checklist')) {
+            node.setAttribute('id', Math.floor(new Date() * Math.random()));
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js:139]: I'm not convinced by this random id. Why not simply always increment ?

File: addons/web_editor/static/src/js/backend/field_html.js [hunk 5]
```diff
@@ -497,6 +499,38 @@ var FieldHtml = basic_fields.DebouncedField.extend(TranslatableFieldMixin, {
...
+            route: '/web_editor/stars',
+            params: {
+                res_model: this.model,
+                res_id: this.res_id,
+                filename: this.name,
+                starsId,
+                rating,
+            },
+        }).then(value => this._setValue(value));
+    },
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/backend/field_html.js:533]: It's a bit sad to have this logic somewhat duplicated with the editable version, though.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js (hunk 3 above)
Comment (Zinston) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:448]: This is explained in the commit message, maybe I need to clarify it then. The feature depends on a callback on the `contentChanged` event, which was not triggered anymore since the new editor was introduced. I fix that here.

File: addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js (hunk 4 above)
Comment (Zinston) [reply addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js:139]: Because precisely it's not simple to always increment and avoid duplication errors. I was able to remove code that was specially dedicated to managing those ids (20 lines, just below this, triggered on every single change in the page), and reuse this for the stars.
Since the code was triggered on change, now that change is linked to an observer, if that code changes the dom it triggers the observer a second time. All in all ridiculously inefficient and complicated for something that can be handled in one line with a random id.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 2 above)
Comment (Zinston) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:1520]: Because this commit fixes the content_changed issue that broke checklists, and this is the dead code that was never triggered. It also includes attachment, focus and blur, which are already handled elsewhere, but I figured it would be a bit ridiculous to just remove content_changed here and create a separate commit for the other three, wouldn't it?

File: addons/web_editor/static/src/js/backend/field_html.js (hunk 5 above)
Comment (Zinston) [reply addons/web_editor/static/src/js/backend/field_html.js:533]: Yes I know, I tried to reduce it as much as possible but couldn't do better than this.
The same was true for checklist by the way, except all that logic was duplicated in python rather than in js :p

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js (hunk 3 above)
Comment (dmo-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:448]: My bad, I did not read it.

File: addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js (hunk 4 above)
Comment (dmo-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js:139]: Why not a uuid then ?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js (hunk 1 above)
Comment (Zinston) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2562]: Yeah you're right. I guess I was blinded by the fact that I didn't support the idea of that part of the feature :p

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+
